src/Viewer/services/decoder/FileManager.js

```diff
@@ -176,6 +190,37 @@ class FileManager {
         this.state.numberOfEvents = this._logEventOffsets.length;
     };
 
+    /**
+     * Given a ref timestamp, find the first log event whose timestamp is 
+     * greater or equal to the ref timestamp.
+     * Since all the log events are naturally sorted w.r.t. the timestamp,
+     * we can use binary search.
+     * Return the index of the log event.
+     * If the log event doesn't exist, the function should return null
+     */
+    _findFirstLogEventWithTimestamp (refTimestamp) {
+        const numberOfEvents = this._logEventOffsets.length;
+        let left = 0;
+        let right = numberOfEvents - 1;
+        let mid;
+
+        while (left <= right) {
+            mid = left + Math.floor((right - left) / 2); // To avoid overflow.
+            if (this._logEventOffsets[mid].timestamp >= refTimestamp) {
+                if (mid === 0 || this._logEventOffsets[mid - 1].timestamp < refTimestamp) {
```

Comment:
I see why you're doing this, but it wasn't immediately obvious when I first read it. I wonder if we could rewrite this algorithm a bit more simply:

* Check if the timestamp is in the range, if not, return early. Although minor, this should speed things up if the timestamp is less than any timestamp in the file.
* Then in the while loop, have three branches: <, ==, >. And you only need to do this special case in the equals branch.

Unless you think this is more confusing, in which case, I guess just apply the first bullet.

Reply:
- I will do the early exit as it increases the readability.
- However, I don't think there's any way we can simplify the inner loop logic. We still have to do the special case check for >, as the given timestamp might not appear in any log event at all.

Reply:
You would be checking if the timestamp appears in the range of log events before the loop, right?

Reply:
Ah, I see what you're saying, my mistake. It would be good if you can add some comments above that condition to explain what it's handling. I know this probably seems like overkill, but it makes it easier to read the code and should only take you a minute since you've already thought about them.